jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java

```diff
@@ -501,6 +502,13 @@ protected void configure(SelectableChannel selectable) throws IOException
             if (sendBufferSize >= 0)
                 setSocketOption(channel, StandardSocketOptions.SO_SNDBUF, sendBufferSize);
         }
+        if (selectable instanceof DatagramChannel)
+        {
+            // QUIC must know the local address, but it is non-null on datagram sockets only if it has been bound,
+            // so implicitly bind to 0.0.0.0:0 when no bind address has been set.
+            if (getBindAddress() == null)
+                setBindAddress(new InetSocketAddress("0.0.0.0", 0));
+        }
```

Comment:
Also, for IPv6 specific runtimes this would be a different address.

The quiche examples bind to these addresses always, just to satisfy things on OSX and BSD.

https://github.com/cloudflare/quiche/blob/21af572b4ae1ce906fbac7b3048409c935e59662/apps/src/client.rs#L73-L79

``` rs
    // Bind to INADDR_ANY or IN6ADDR_ANY depending on the IP family of the
    // server address. This is needed on macOS and BSD variants that don't
    // support binding to IN6ADDR_ANY for both v4 and v6.
    let bind_addr = match peer_addr {
        std::net::SocketAddr::V4(_) => format!("0.0.0.0:{}", args.source_port),
        std::net::SocketAddr::V6(_) => format!("[::]:{}", args.source_port),
    };
```

Reply:
My understanding is that QUIC allows `Connection Migration`. This was only implemented in Quiche 0.15.0 and required a quiche API change ([see here](https://github.com/cloudflare/quiche/releases/tag/0.15.0)).

This means migrating a connection from one path (e.g.:IP1 <=> IP2) to another path (e.g.: IP3 <=> IP2) now requires quiche to know about the source and destination IP addresses. AFAICT the only way to get the right IP and port is to bind the local address otherwise querying the socket always returns null.

Is there another way, or did I miss something, or could we move that code somewhere else? Maybe, I'm open to discussing this.

Reply:
After a discussion with @sbordet we figured this binding is needed only the client needs this extra one, and it can figure the IP address if `null` is passed to `bind()` so this was moved to the quic client connector configurator.